## Ejercicio en casa 2
##### Dataset de Netflix (Contenidos)
https://www.kaggle.com/datasets/swatikhedekar/exploratory-data-analysis-on-netflix-data



In [2]:
import pandas as pd
import matplotlib.pyplot as plt
df = pd.read_csv("netflix_titles_2021.csv")
df.head(2)

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   

                                         listed_in  \
0                                    Documentaries   
1  International TV Shows, TV Dramas, TV Mysteries   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...

# Ejercicio en Casa:

Para este ejercicio debes realizar los siguientes pasos de manera anticipada los siguientes procesos:
- Cargar el conjunto de datos desde un archivo CSV utilizando pandas.
- Explorar y familiarizarse con la estructura del conjunto de datos.
- Calcular estadísticas descriptivas, como el promedio, la mediana y la desviación estándar, de variables relevantes del conjunto de datos general.

#### Acciones por ejecutar

1. Explorar inconsistencias
2. Separar contenidos con duración en minutos y temporadas
3. Homogenizar para detectar preguntas como: Tiempo promedio, temporadas promedio.
4. Relación entre listed_in y duration (separado mins y temporadas)
5. Frecuencia de distribución de los paises
6. Actores principales por pais
7. Relación entre tipo y duración
8. Relación entre listed_in y tipo

In [3]:
# Exploración básica inicial
df.shape

(8807, 12)

In [5]:
# Características disponibles
df.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'],
      dtype='object')

In [6]:
# Tipos de cada columna
df.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

In [7]:
# información general por columna
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [8]:
# Exploración general
for col in df.columns:
    print("Columna: ", col, "[", df[col].nunique(), "]")
    print(df[col].sort_values().unique())

Columna:  show_id [ 8807 ]
['s1' 's10' 's100' ... 's997' 's998' 's999']
Columna:  type [ 2 ]
['Movie' 'TV Show']
Columna:  title [ 8807 ]
['#Alive' '#AnneFrank - Parallel Stories' '#FriendButMarried' ... '마녀사냥'
 '반드시 잡는다' '최강전사 미니특공대 : 영웅의 탄생']
Columna:  director [ 4528 ]
['A. L. Vijay' 'A. Raajdheep' 'A. Salaam' ... 'Ömer Faruk Sorak'
 'Şenol Sönmez' nan]
Columna:  cast [ 7692 ]
["'Najite Dede, Jude Chukwuka, Taiwo Arimoro, Odenike Odetola, Funmi Eko, Keppy Ekpenyong"
 '4Minute, B1A4, BtoB, ELSIE, EXID, EXO, Got7, INFINITE, KARA, Shinee, Sistar, VIXX, Nine Muses, BTS, Secret, Topp Dogg'
 '50 Cent, Ryan Phillippe, Bruce Willis, Rory Markham, Jenna Dewan, Brett Granstaff, Randy Couture, Susie Abromeit, Ron Turner, James Remar'
 ...
 'Şükrü Özyıldız, Aslı Enver, Şenay Gürler, Başak Parlak, Mahir Günşiray, Hakan Boyav, Hakan Gerçek, Berrak Kuş, Gamze Süner Atay, Mehmet Esen'
 'Ṣọpẹ́ Dìrísù, Wunmi Mosaku, Matt Smith, Malaika Wakoli-Abigaba' nan]
Columna:  country [ 748 ]
[', France, Algeri

In [9]:
# Exploración
## Identificación de NaN
for col in df.columns:
    print("Nan x columna [", col, "]:", df[col].isnull().sum())    

Nan x columna [ show_id ]: 0
Nan x columna [ type ]: 0
Nan x columna [ title ]: 0
Nan x columna [ director ]: 2634
Nan x columna [ cast ]: 825
Nan x columna [ country ]: 831
Nan x columna [ date_added ]: 10
Nan x columna [ release_year ]: 0
Nan x columna [ rating ]: 4
Nan x columna [ duration ]: 3
Nan x columna [ listed_in ]: 0
Nan x columna [ description ]: 0


In [97]:
df = pd.read_csv("netflix_titles_2021.csv")
df.head(2)

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   

                                         listed_in  \
0                                    Documentaries   
1  International TV Shows, TV Dramas, TV Mysteries   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...

### 2. Separar contenidos con duración en minutos y temporadas (duration)

#### Separar usando expresión regular dígito + palabra
#### Opción 1

In [98]:
# Opción 1
df[['duration_unit', 'time_unit']] = df['duration'].str.extract(r'(\d+) (\w+)')
df.head(2)

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   

                                         listed_in  \
0                                    Documentaries   
1  International TV Shows, TV Dramas, TV Mysteries   

                                         description duration_unit time_unit  
0  As her father nears the end of his life, filmm...            90       min  
1  After crossing paths at a party, a Cape Town t...             2   Seasons

#### Separar usando split de espacio
#### Opción 2

In [17]:
## https://pandas.pydata.org/docs/user_guide/text.html
splitted_list = df["duration"].str.split(" ")
splitted_list
df["duration_unit 2"] = pd.DataFrame(splitted_list)["duration"].str.get(0)
df["time_unit 2"] = pd.DataFrame(splitted_list)["duration"].str.get(1)
df.head(2)

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   

                                         listed_in  \
0                                    Documentaries   
1  International TV Shows, TV Dramas, TV Mysteries   

                                         description duration_unit time_unit  \
0  As her father nears the end of his life, filmm...            90       min   
1  After crossing paths at a party, a Cape Town t...             2   Seasons   

  Numero    Unidad duration_unit 2 time_unit 2  
0     90       min              90         min  
1      2   Seasons               2     Seasons

#### Opción 3 By Jazmin

In [18]:
df['Numero']=df['duration'].str.extract('(\d+)') # separa la parte numérica
df['Unidad']=df['duration'].str.extract('(\D+)') # separa por Unidades
df.head(2)

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   

                                         listed_in  \
0                                    Documentaries   
1  International TV Shows, TV Dramas, TV Mysteries   

                                         description duration_unit time_unit  \
0  As her father nears the end of his life, filmm...            90       min   
1  After crossing paths at a party, a Cape Town t...             2   Seasons   

  Numero    Unidad duration_unit 2 time_unit 2  
0     90       min              90         min  
1      2   Seasons               2     Seasons

#### Opcion 4 - Oscar 

In [19]:
df.insert(df.columns.get_loc('duration')+1, 'unit', '')
df[['duration', 'unit']] = df['duration'].str.split(' ', expand=True)
df['unit'] = df['unit'].str.replace('Seasons', 'Season')

#### Opcion 5 - Manuel

In [27]:
df[['Tiempo','UnidadTiempo']] = df['duration'].str.split(' ', expand=True)
df['UnidadTiempo'] = df['UnidadTiempo'].replace('Seasons', 'Season')
df['Tiempo'] = pd.to_numeric(df['Tiempo'], errors='coerce')
df.head(2)

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   

                                         listed_in  \
0                                    Documentaries   
1  International TV Shows, TV Dramas, TV Mysteries   

                                         description duration_unit time_unit  \
0  As her father nears the end of his life, filmm...            90       min   
1  After crossing paths at a party, a Cape Town t...             2   Seasons   

   Tiempo UnidadTiempo  
0    90.0          min  
1     2.0       Season

#### Opcion 5 - Ricardo

In [35]:
df[['Valor', 'unidad']] = df['duration'].str.split(expand=True)
df.head(2)

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   

                                         listed_in  \
0                                    Documentaries   
1  International TV Shows, TV Dramas, TV Mysteries   

                                         description duration_unit time_unit  \
0  As her father nears the end of his life, filmm...            90       min   
1  After crossing paths at a party, a Cape Town t...             2   Seasons   

  Valor   unidad  
0    90      min  
1     2  Seasons

#### Opcion 6 - Joan

In [38]:
df2 = df[['duration','units']] = df.duration.str.split(' ',expand=True)

ValueError: Columns must be same length as key

In [100]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   show_id        8807 non-null   object
 1   type           8807 non-null   object
 2   title          8807 non-null   object
 3   director       6173 non-null   object
 4   cast           7982 non-null   object
 5   country        7976 non-null   object
 6   date_added     8797 non-null   object
 7   release_year   8807 non-null   int64 
 8   rating         8803 non-null   object
 9   duration       8804 non-null   object
 10  listed_in      8807 non-null   object
 11  description    8807 non-null   object
 12  duration_unit  8804 non-null   object
 13  time_unit      8804 non-null   object
dtypes: int64(1), object(13)
memory usage: 963.4+ KB


In [33]:
# df.dropna(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   show_id        8807 non-null   object
 1   type           8807 non-null   object
 2   title          8807 non-null   object
 3   director       6173 non-null   object
 4   cast           7982 non-null   object
 5   country        7976 non-null   object
 6   date_added     8797 non-null   object
 7   release_year   8807 non-null   int64 
 8   rating         8803 non-null   object
 9   duration       8804 non-null   object
 10  listed_in      8807 non-null   object
 11  description    8807 non-null   object
 12  duration_unit  8804 non-null   object
 13  time_unit      8804 non-null   object
dtypes: int64(1), object(13)
memory usage: 963.4+ KB


### 3. Homogenizar para detectar preguntas como:

#### - Tiempo promedio | - Temporadas promedio.

#### ¿Qué situaciones se estén presentando en este caso?

1. Se logra dividir pero tenemos peras y manzanas, por qué? porque hay una dependencia de tipo... ("movies", "TV shows")
2. Las medidas están viciadas, la media no es la media... y así

In [64]:
df["duration_unit"].dropna().astype("int16").describe()

count    8804.000000
mean       69.846888
std        50.814828
min         1.000000
25%         2.000000
50%        88.000000
75%       106.000000
max       312.000000
Name: duration_unit, dtype: float64

In [101]:
df.dropna(subset=["duration_unit"], inplace=True)
df["duration_unit"] = df["duration_unit"].astype("int16")
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8804 entries, 0 to 8806
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   show_id        8804 non-null   object
 1   type           8804 non-null   object
 2   title          8804 non-null   object
 3   director       6170 non-null   object
 4   cast           7979 non-null   object
 5   country        7973 non-null   object
 6   date_added     8794 non-null   object
 7   release_year   8804 non-null   int64 
 8   rating         8800 non-null   object
 9   duration       8804 non-null   object
 10  listed_in      8804 non-null   object
 11  description    8804 non-null   object
 12  duration_unit  8804 non-null   int16 
 13  time_unit      8804 non-null   object
dtypes: int16(1), int64(1), object(12)
memory usage: 980.1+ KB


* Nótese que los valores se calculan sobre 8804 registros tras el dropna()

In [102]:
## Se observa que no se han eliminado los na y siguyen siendo 8807 registros.
print(df["duration_unit"].isnull().sum()," :|: ",df["duration_unit"].isna().sum())

0  :|:  0


In [103]:
#type(df["duration_unit"].dropna())
#type(df["duration_unit"].dropna(inplace=True))

## Se observa que no se han eliminado los na y siguyen siendo 8807 registros.
print(df["duration_unit"].isnull().sum()," :|: ",df["duration_unit"].isna().sum())

0  :|:  0


In [54]:
df = df.dropna(subset=["duration_unit"])
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8804 entries, 0 to 8806
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   show_id        8804 non-null   object
 1   type           8804 non-null   object
 2   title          8804 non-null   object
 3   director       6170 non-null   object
 4   cast           7979 non-null   object
 5   country        7973 non-null   object
 6   date_added     8794 non-null   object
 7   release_year   8804 non-null   int64 
 8   rating         8800 non-null   object
 9   duration       8804 non-null   object
 10  listed_in      8804 non-null   object
 11  description    8804 non-null   object
 12  duration_unit  8804 non-null   int16 
 13  time_unit      8804 non-null   object
 14  Valor          8804 non-null   object
 15  unidad         8804 non-null   object
 16  units          8804 non-null   object
dtypes: int16(1), int64(1), object(15)
memory usage: 1.2+ MB


#### Crear una vista de "movies"

Recuerde que la vista y la copia son diferentes la copia es valor, y la vista es referencia

In [104]:
df_movies = df[df["type"] == "Movie"]
df_movies.head(2)

show_id   type                             title  \
0      s1  Movie              Dick Johnson Is Dead   
6      s7  Movie  My Little Pony: A New Generation   

                        director  \
0                Kirsten Johnson   
6  Robert Cullen, José Luis Ucha   

                                                cast        country  \
0                                                NaN  United States   
6  Vanessa Hudgens, Kimiko Glenn, James Marsden, ...            NaN   

           date_added  release_year rating duration                 listed_in  \
0  September 25, 2021          2020  PG-13   90 min             Documentaries   
6  September 24, 2021          2021     PG   91 min  Children & Family Movies   

                                         description  duration_unit time_unit  
0  As her father nears the end of his life, filmm...             90       min  
6  Equestria's divided. But a bright-eyed hero be...             91       min

#### Crear una vista de "TV Show"

La vista sería válida de cara a realizar agrupaciones u otras operaciones descendentes

In [59]:
df_tv_show = df[df["type"] =="TV Show"]
df_tv_show.head(2)

show_id     type          title         director  \
1      s2  TV Show  Blood & Water              NaN   
2      s3  TV Show      Ganglands  Julien Leclercq   

                                                cast       country  \
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...  South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...           NaN   

           date_added  release_year rating duration  \
1  September 24, 2021          2021  TV-MA        2   
2  September 24, 2021          2021  TV-MA        1   

                                           listed_in  \
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   

                                         description  duration_unit time_unit  \
1  After crossing paths at a party, a Cape Town t...              2   Seasons   
2  To protect his family from a powerful drug lor...              1    Season   

  Valor   unidad    units  
1     2  Seasons  Seasons  
2     1   Season   Season

### Operando sobre las vistas

In [62]:
# Cálculo de la media en la columna de duración para las series (TV Show)
serie_duration_mean = df_tv_show.duration_unit.dropna().astype("int16").mean()
print("La medía de duración de una serie es: ", serie_duration_mean,"temporadas", type(serie_duration_mean))

La medía de duración de una serie es:  1.764947683109118 temporadas <class 'numpy.float64'>


In [63]:
# Cálculo de la media en la columna de duración para las películas (Movies)
movie_duration_mean = df_movies.duration_unit.dropna().astype("int16").mean()
print("La medía de duración de una pelicula es: ", movie_duration_mean,"minutos")

La medía de duración de una pelicula es:  99.57718668407311 minutos


In [65]:
df_tv_show.duration_unit.dropna().astype("int16").describe()

count    2676.000000
mean        1.764948
std         1.582752
min         1.000000
25%         1.000000
50%         1.000000
75%         2.000000
max        17.000000
Name: duration_unit, dtype: float64

In [66]:
df_movies.duration_unit.dropna().astype("int16").describe()

count    6128.000000
mean       99.577187
std        28.290593
min         3.000000
25%        87.000000
50%        98.000000
75%       114.000000
max       312.000000
Name: duration_unit, dtype: float64

In [67]:
df_movies = df_movies[ df_movies["duration_unit"].isnull() == False ]
len(df_movies)

6128

# Usando copias

#### Crear una *copia* de "TV Show"

Veamos la diferencia en la media respecto al "combinado"

In [68]:
# las "tv_show" son de tipo seasons
df_tv_show_copy = df.copy()
df_tv_show_copy = df_tv_show_copy[df_tv_show_copy["type"] =="TV Show"]
df_tv_show_copy.dropna(inplace=True)
df_tv_show_copy["duration_unit"].astype('int16').mean()

1.9047619047619047

#### Crear una *copia* de "movies"

Veamos la diferencia en la media respecto al "combinado"

In [105]:
df_movies_copy = df.copy()
df_movies_copy = df_movies_copy[df_movies_copy["type"] =="Movie"]
df_movies_copy.dropna(inplace=True)
df_movies_copy["duration_unit"].astype('int16').mean()

102.70144648023144

### 4. Relación entre listed_in y duration (separado mins y temporadas)

In [106]:
df_movies["listed_in"].nunique()

278

In [107]:
df_movies["duration_unit"].nunique()

205

In [108]:
df_movies["listed_in"]

0                                          Documentaries
6                               Children & Family Movies
7       Dramas, Independent Movies, International Movies
9                                       Comedies, Dramas
12                          Dramas, International Movies
                              ...                       
8801             Dramas, International Movies, Thrillers
8802                      Cult Movies, Dramas, Thrillers
8804                             Comedies, Horror Movies
8805                  Children & Family Movies, Comedies
8806      Dramas, International Movies, Music & Musicals
Name: listed_in, Length: 6128, dtype: object

In [73]:
listas = df_movies["listed_in"].str.split(",")
listas

0                                         [Documentaries]
6                              [Children & Family Movies]
7       [Dramas,  Independent Movies,  International M...
9                                     [Comedies,  Dramas]
12                        [Dramas,  International Movies]
                              ...                        
8801          [Dramas,  International Movies,  Thrillers]
8802                   [Cult Movies,  Dramas,  Thrillers]
8804                           [Comedies,  Horror Movies]
8805                [Children & Family Movies,  Comedies]
8806    [Dramas,  International Movies,  Music & Music...
Name: listed_in, Length: 6128, dtype: object

### Opciones para el conteo de categorías

#### Opción 1 - a mano ✊

In [82]:
# Paso 1: Separar todas las categorías en una lista de listas
categorias_listas = df_movies["listed_in"].str.split(",").apply(lambda lista: [categoria.strip() for categoria in lista])
categorias_listas

0                                         [Documentaries]
6                              [Children & Family Movies]
7       [Dramas, Independent Movies, International Mov...
9                                      [Comedies, Dramas]
12                         [Dramas, International Movies]
                              ...                        
8801            [Dramas, International Movies, Thrillers]
8802                     [Cult Movies, Dramas, Thrillers]
8804                            [Comedies, Horror Movies]
8805                 [Children & Family Movies, Comedies]
8806     [Dramas, International Movies, Music & Musicals]
Name: listed_in, Length: 6128, dtype: object

In [77]:
# Paso 2: Unir todas las listas en una sola lista
todas_categorias = [categoria for lista in categorias_listas for categoria in lista]
todas_categorias

['Documentaries',
 'Children & Family Movies',
 'Dramas',
 'Independent Movies',
 'International Movies',
 'Comedies',
 'Dramas',
 'Dramas',
 'International Movies',
 'Children & Family Movies',
 'Comedies',
 'Documentaries',
 'International Movies',
 'Thrillers',
 'Comedies',
 'International Movies',
 'Children & Family Movies',
 'Comedies',
 'International Movies',
 'Romantic Movies',
 'Comedies',
 'International Movies',
 'Music & Musicals',
 'Comedies',
 'Horror Movies',
 'Sci-Fi & Fantasy',
 'Thrillers',
 'Dramas',
 'Independent Movies',
 'International Movies',
 'Dramas',
 'International Movies',
 'Thrillers',
 'Action & Adventure',
 'Dramas',
 'International Movies',
 'Action & Adventure',
 'Dramas',
 'Action & Adventure',
 'Classic Movies',
 'Dramas',
 'Dramas',
 'Horror Movies',
 'Thrillers',
 'Action & Adventure',
 'Horror Movies',
 'Thrillers',
 'Action & Adventure',
 'Horror Movies',
 'Thrillers',
 'Documentaries',
 'Action & Adventure',
 'Dramas',
 'Thrillers',
 'Action & 

In [80]:
# Paso 3: Obtener los elementos únicos (usando un conjunto ---> no soporta repetidos)
elementos_unicos = set(todas_categorias)
elementos_unicos

{'Action & Adventure',
 'Anime Features',
 'Children & Family Movies',
 'Classic Movies',
 'Comedies',
 'Cult Movies',
 'Documentaries',
 'Dramas',
 'Faith & Spirituality',
 'Horror Movies',
 'Independent Movies',
 'International Movies',
 'LGBTQ Movies',
 'Movies',
 'Music & Musicals',
 'Romantic Movies',
 'Sci-Fi & Fantasy',
 'Sports Movies',
 'Stand-Up Comedy',
 'Thrillers'}

In [81]:
print(elementos_unicos, len(elementos_unicos))

{'Horror Movies', 'LGBTQ Movies', 'Comedies', 'Romantic Movies', 'Independent Movies', 'Movies', 'Anime Features', 'Stand-Up Comedy', 'Sports Movies', 'Action & Adventure', 'Dramas', 'Children & Family Movies', 'Cult Movies', 'Faith & Spirituality', 'Music & Musicals', 'Sci-Fi & Fantasy', 'Classic Movies', 'Documentaries', 'Thrillers', 'International Movies'} 20


#### Opción 2 - con la magia de la librería 🧙‍♂️

In [86]:
# Paso 1: Separar todas las categorías en una única lista
listado_categorias = categorias_listas.explode()
listado_categorias

0                  Documentaries
6       Children & Family Movies
7                         Dramas
7             Independent Movies
7           International Movies
                  ...           
8805    Children & Family Movies
8805                    Comedies
8806                      Dramas
8806        International Movies
8806            Music & Musicals
Name: listed_in, Length: 13187, dtype: object

In [87]:
# Paso 2: realizar el conteo
cantidad_por_categoria = listado_categorias.value_counts()

In [89]:
# Paso 3: visualizar
cantidad_por_categoria

International Movies        2752
Dramas                      2427
Comedies                    1674
Documentaries                869
Action & Adventure           859
Independent Movies           756
Children & Family Movies     641
Romantic Movies              616
Thrillers                    577
Music & Musicals             375
Horror Movies                357
Stand-Up Comedy              343
Sci-Fi & Fantasy             243
Sports Movies                219
Classic Movies               116
LGBTQ Movies                 102
Cult Movies                   71
Anime Features                71
Faith & Spirituality          65
Movies                        54
Name: listed_in, dtype: int64

In [109]:
df_movies_by_listin = df_movies[ df_movies["listed_in"].str.contains("Music & Musicals") ]
df_movies_by_listin.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 375 entries, 26 to 8806
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   show_id        375 non-null    object
 1   type           375 non-null    object
 2   title          375 non-null    object
 3   director       358 non-null    object
 4   cast           340 non-null    object
 5   country        328 non-null    object
 6   date_added     375 non-null    object
 7   release_year   375 non-null    int64 
 8   rating         375 non-null    object
 9   duration       375 non-null    object
 10  listed_in      375 non-null    object
 11  description    375 non-null    object
 12  duration_unit  375 non-null    int16 
 13  time_unit      375 non-null    object
dtypes: int16(1), int64(1), object(12)
memory usage: 41.7+ KB


In [110]:
print("El tiempo promedio de una pelicula en la categoría es:" , int(df_movies_by_listin["duration_unit"].dropna().astype('int16').mean()), "minutos")

El tiempo promedio de una pelicula en la categoría es: 106 minutos


## 5. Frecuencia de distribución de los paises

In [115]:
df.country.isnull().sum()

831

In [117]:
df.dropna(subset=["country"], inplace=True)
df.country.isnull().sum()

0

In [119]:
# Paso 1: Separar todos los paises en una única lista
listado_paises = df["country"].str.split(", ").explode()
listado_paises

0       United States
1        South Africa
4               India
7       United States
7               Ghana
            ...      
8801           Jordan
8802    United States
8804    United States
8805    United States
8806            India
Name: country, Length: 10014, dtype: object

In [120]:
# Paso 2: realizar el conteo
cantidad_por_pais = listado_paises.value_counts()

# Paso 3: visualizar
cantidad_por_pais

United States     3689
India             1046
United Kingdom     804
Canada             445
France             393
                  ... 
Mongolia             1
Albania              1
Samoa                1
Palestine            1
Bahamas              1
Name: country, Length: 127, dtype: int64

### 6. Actores principales por pais

In [121]:
# Paso 1: Separar todos los actores en una única lista
Listado_actores = df["cast"].str.split(", ").explode()

# Paso 2: realizar el conteo
cantidad_actor = Listado_actores.value_counts()

# Paso 3: visualizar
cantidad_actor

Anupam Kher          43
Shah Rukh Khan       34
Naseeruddin Shah     31
Akshay Kumar         30
Om Puri              30
                     ..
Kunika                1
Da Peng               1
Marta Marco           1
Ruangroj Prachaya     1
Eve Arden             1
Name: cast, Length: 34321, dtype: int64

In [122]:
Listado_paises = Listado_paises.str.replace(",","")
listado_paises_unicos = list(set(Listado_paises))

In [123]:
listado_paises_unicos.sort()
listado_paises_unicos

['',
 'Afghanistan',
 'Albania',
 'Algeria',
 'Angola',
 'Argentina',
 'Armenia',
 'Australia',
 'Austria',
 'Azerbaijan',
 'Bahamas',
 'Bangladesh',
 'Belarus',
 'Belgium',
 'Bermuda',
 'Botswana',
 'Brazil',
 'Bulgaria',
 'Burkina Faso',
 'Cambodia',
 'Cameroon',
 'Canada',
 'Cayman Islands',
 'Chile',
 'China',
 'Colombia',
 'Croatia',
 'Cuba',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'Dominican Republic',
 'East Germany',
 'Ecuador',
 'Egypt',
 'Ethiopia',
 'Finland',
 'France',
 'Georgia',
 'Germany',
 'Ghana',
 'Greece',
 'Guatemala',
 'Hong Kong',
 'Hungary',
 'Iceland',
 'India',
 'Indonesia',
 'Iran',
 'Iraq',
 'Ireland',
 'Israel',
 'Italy',
 'Jamaica',
 'Japan',
 'Jordan',
 'Kazakhstan',
 'Kenya',
 'Kuwait',
 'Latvia',
 'Lebanon',
 'Liechtenstein',
 'Lithuania',
 'Luxembourg',
 'Malawi',
 'Malaysia',
 'Malta',
 'Mauritius',
 'Mexico',
 'Mongolia',
 'Montenegro',
 'Morocco',
 'Mozambique',
 'Namibia',
 'Nepal',
 'Netherlands',
 'New Zealand',
 'Nicaragua',
 'Nigeria',
 'Nor

In [ ]:
### ------------------

In [124]:
import pandas as pd
import matplotlib.pyplot as plt
from collections import Counter
df = pd.read_csv("netflix_titles_2021.csv")
df.head(2)

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   

                                         listed_in  \
0                                    Documentaries   
1  International TV Shows, TV Dramas, TV Mysteries   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...

In [126]:
import pandas as pd
from collections import Counter

# Ejemplo de DataFrame
data = {
    'show_id': [1, 2, 3, 4, 5],
    'type': ['Movie', 'Movie', 'Movie', 'TV Show', 'Movie'],
    'title': ['Movie 1', 'Movie 2', 'Movie 3', 'TV Show 1', 'Movie 4'],
    'director': ['Director 1', 'Director 2', 'Director 3', 'Director 4', 'Director 5'],
    'cast': ['Actor A, Actor B', 'Actor B, Actor C', 'Actor A, Actor C', 'Actor D, Actor E', 'Actor B, Actor E'],
    'country': ['Country 1, Country 2', 'Country 2', 'Country 1', 'Country 3', 'Country 2, Country 3']
}

df = pd.DataFrame(data)

# Paso 1: Convertir las celdas separadas por comas en registros individuales
df['country'] = df['country'].str.split(', ')

# Paso 2: Utilizar explode() en la columna "country"
df_exploded = df.explode('country')

# Paso 3: Agrupar por país y contar los actores
grouped = df_exploded.groupby('country')['cast'].apply(lambda x: ','.join(x)).str.split(',')
actor_counts_by_country = grouped.apply(lambda x: Counter(x))

# Paso 4: Obtener el actor con más repeticiones por país
most_common_actors_by_country = actor_counts_by_country.apply(lambda x: x.most_common(1)[0][0])

print(most_common_actors_by_country)

country
Country 1     Actor A
Country 2     Actor B
Country 3     Actor E
Name: cast, dtype: object
